### PR TITLE
Remove config/addon-sample.config.php file in favor of per-addon configuration files

### LINKED ISSUE #10188: Problem with LDAP Auth and Active Directory "Add on not configured or missing php-ldap extension"
Hello everyone!

I seem to have a problem connecting friendica with my Active directory for user authentication. I'm using the docker version `friendica:latest` with a  `mariadb:latest` database.

I have:

- enabled the addon through the admin page
- Filled addon.config.php with the LDAP info as suggested by the configuration

> return [
>         'ldapauth' => [
>                 'ldap_server' => '192.168.x.x',
>                 'ldap_binddn' => ''CN=ldapread,OU=ITUSERS,DC=mydomain,DC=tld',
>                 'ldap_bindpw' => 'verysecretpassword',
>                 'ldap_searchdn' => 'OU=SUSERS,DC=mydomain,DC=tld',
>                 'ldap_userattr' => 'sAMAccountName',
>                 'ldap_group' => ''CN=friends,OU=SERVICES,OU=GROUPS,DC=mydomain,DC=tld',
>                 'ldap_autocreateaccount' => true,
>                 'ldap_autocreateaccount_emailattribute' => 'mail',
>                 'ldap_autocreateaccount_nameattribute' => 'sAMAccountName',
>         ],
> ];

I also tried using the hostname for the AD.

- I checked the addon table in the database and the addon is installed = 1
- I made sure the php-ldap extension is present (in case the docker image had this missing)

After all those I still get this log messages:

```
2021-05-01T10:06:18Z index [ERROR]: Addon not configured or missing php-ldap extension {"extension_loaded":true,"server":""} - {"file":"ldapauth.php","line":97,"function":"ldapauth_authenticate","uid":"261d17","process_id":45}
2021-05-01T10:06:18Z index [WARNING]: authenticate: failed login attempt {"action":"login","username":"user@mydomain.ltd","ip":"192.168.42.20"} - {"file":"Authentication.php","line":271,"function":"withPassword","uid":"261d17","process_id":45}
```

I really hope I havved missed anything profound like Active directory not being supported by the plugin. I don't use secure conection with the AD. I also am not sure if I should use a hostname, an IP or an IP:PORT for `ldap_server` but I did try all f those.

Nevertheless I thing the log error shows that the addon doesn't get recognized as `Enabled` so it won't even try to connect.
Unfortunately I couldn't find any similar issues posted.

Thanks in advance for your time. Any nudge to the right direction will be much appreciated! :)

### PR DESCRIPTION
**Pull request by MrPetovan**
Address https://github.com/friendica/friendica/issues/12135#issuecomment-1321098500

I got rid of the unique `config/addon-sample.config.php` file in favor of per-addon configuration file given https://github.com/friendica/friendica/issues/12135#issuecomment-1321098500 where the documentation assumes some pre-existing knowledge about the config files that shouldn't be necessary. This change won't have any effect on existing installs as the existing `config/addon.config.php` file will still be loaded as part of the `config/` folder. We just don't advertise its existence in the documentation anymore.

### CODE REVIEW

**Comment by nupplaphil** [comment]
I think it's now more confusing than before and I'm not sure what regressions are now possible because of the double call (because manually reloading the config inside `App`, it's not even sure if we catch every occurrence - like for console commands).

There are currently 5 (ordered) kind of config sources

- `SOURCE_STATIC` => values from `default.config.php` and `settings.config.php`
- `SOURCE_FILE` => any  value  from a custom `*.config.php` file from `config/` or `addon/*/config/`
- `SOURCE_DB` =>  any  value from the `config` table
- `SOURCE_ENV` => any value from an `Environment` variable
- `SOURCE_FIX` => any "fixed" value, a config could set manually in the code (currently not used)

This order is strict and ensures the order `static` < `file` < `db` < `env`

Now the other dimension is the timing and I would like to introduce the `static` folder in the addons as well (instead of reloading the config again), so we would have:

- Loading `static/` values
- Loading `addon/*/static/` values
- Loading `config/` values
- Loading `addon/*/config/` values
- Loading the DB
- Loading the Environment

For me, it's better understandable than reloading the core config again :-)

[edit] Therefor I'd say let's move the `Core\Hook::callAll('load_config');` inside the Config class to avoid this reloading ...

**Comment by nupplaphil** [comment]
But I think I get it ... örks ...
Loading the Addons static files would needed to be after loading the db, but before loading the config, which isn't possible because the DB config is loaded through the "normal" config loading .. omg ..

**Comment by MrPetovan** [comment]
I got to the same point as you and got stuck there.

From the priority list apparently I didn't need to reload the env since it would be still be priority over the file that are loaded again.

**Comment by nupplaphil** [comment]
Ah I think this should work:

- create a `static/` folder for each addon with default values
- Each addon with `static` folder should create a `addon_load_static_config()` function and load it's static files with the parameter `SOURCE_STATIC` (=> so it won't override the `config/` values anymore)
- add a new `Core\Hook::callAll('load_static_config', $loader);` call at `App`

**Comment by MrPetovan** [comment]
Oooh, shiny. However the only thing we need is to set `SOURCE_STATIC` in the current hook and we're good. I'll get on it.

**Comment by MrPetovan** [comment]
Here you go.

File: doc/Config.md [hunk 1]
```diff
@@ -37,8 +37,8 @@ The `config` directory holds key configuration files and can have different conf
 All of them have to end with `.config.php` and must not include `-sample` in their name.
 
 Some examples of common known configuration files:
-- `local.config.php` holds the current node custom configuration.
-- `addon.config.php` is optional and holds the custom configuration for specific addons.
+- `local.config.php` holds the base node custom configuration.
+- Any other file in this folder is meant for addon configuration.
```
Comment (nupplaphil) [inline doc/Config.md]: Any other file is loaded as well in no order ..
have a look at all the configs at https://github.com/friendica/docker/tree/stable/.config
That was one of the main goals for the `Config` changes in the past :)

**Review by nupplaphil — CHANGES_REQUESTED** [review]

File: doc/Config.md (hunk 1 above)
Comment (MrPetovan) [reply doc/Config.md]: No order?

Comment (nupplaphil) [reply doc/Config.md]: ok sry, too much DEnglisch *gg*

The only order inside `config/` is the "standard" directory order, means that `local.config.php` isn't even necessary anymore. Any config file will be loaded, regardless it's naming (except the postfix `config.php`)

Comment (MrPetovan) [reply doc/Config.md]: It's true, but the `local.config.php` file is tied to the `local-sample.config.php` we provide, and it's also the name of the file we create during automatic install. So it isn't technically necessary, but it still serves a couple purposes.

Comment (nupplaphil) [reply doc/Config.md]:
```suggestion
- Any other file in this folder is meant for additional configuration (e.g. for addons).
```

**Pull request merged**